sublime_jedi/completion.py

```diff
@@ -146,61 +154,69 @@ def on_query_completions(self, prefix, locations):
         logger.info('JEDI completion triggered.')
 
         settings = get_settings(self.view)
-        only_jedi_completion = (
-            settings['sublime_completions_visibility'] in ('default', 'jedi')
-        )
-
-        previous_char = self.view.substr(locations[0] - 1)
         if settings['only_complete_after_regex']:
+            previous_char = self.view.substr(locations[0] - 1)
             if not re.match(settings['only_complete_after_regex'], previous_char):
                 return False
 
-        cplns = ask_daemon_with_timeout(
-            self.view,
-            'autocomplete',
-            location=locations[0],
-            timeout=settings['completion_timeout']
-        )
-
-        logger.info("Completion completed.")
-
-        cplns = [tuple(x) for x in self._sort_completions(cplns)]
-        logger.debug("Completions: {0}".format(cplns))
-
-        # disabled due to can't reproduce
-        # self._fix_tab_completion_issue()
-
-        if only_jedi_completion:
-            return cplns, PLUGIN_ONLY_COMPLETION
-        return cplns
+        with self._lock:
+            if self._last_location != locations[0]:
+                self._last_location = locations[0]
+                ask_daemon(
+                    self.view,
+                    self._receive_completions,
+                    'autocomplete',
+                    location=locations[0],
+                )
+                return [], PLUGIN_ONLY_COMPLETION
+
+            if self._last_location == locations[0] and self._completions:
+                self._last_location = None
+                return [
+                    tuple(x)
+                    for x in self._sort_completions(self._completions)
+                ]
+
+    def _receive_completions(self, view, completions):
+        if not completions:
+            return
+
+        logger.debug("Completions: {0}".format(completions))
+
+        with self._lock:
+            self._previous_completions = self._completions
+            self._completions = completions
+
+        if (completions and
+                not view.is_auto_complete_visible() or
+                not self._is_completions_subset()):
```

Comment:
would I see completion in next case:

I'm typing "import a"
i'm hitting "completion" keys - I would receive, something like, 'abc, etc.'
I do not see my module due to list quite big,
I'm continue typing "import ab"
and I'm hitting "completion" keys again - would I receive a smaller list of suggestions?

Reply:
This happens also in `master` so, I guess is a problem of jedi?

This is mostly to prevent flickering. If you type it hides and shows the completion popup again very fast, but sublime does filtering automatically so if the new completions are a subset of the previous ones and the popup is being displayed I will do nothing. Let sublime filter.

There is an error in this condition anyway, I'm fixing it.